File: neetcode/blind75/sliding_window/buy_and_sell_crypto.py
# https://neetcode.io/problems/buy-and-sell-crypto


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0

        # window_size = 1

        profit = 0

        # while window_size < len(prices):
        #     for i in range(len(prices) - window_size):
        #         j = window_size + i
        #         profit = max(profit, prices[j] - prices[i])

        #     window_size += 1

        # return profit

        buy = prices[0]
        for price in prices:
            if price < buy:
                buy = price

            profit = max(profit, price - buy)
        return profit

File: neetcode/neet250_JS/Sliding-Window/buy-and-sell-crypto.js
// https://neetcode.io/problems/buy-and-sell-crypto

/*
create max to track max profit
buy = prices[0]

for (right = 0; right < prices.length; right ++)
    max = math.max(max, prices[i] - buy)

    if the price at right < buy, buy = prices[right]

return max

- Time: O(n)
- Space: O(1)

*/

class Solution {
    /**
     * @param {number[]} prices
     * @return {number}
     */
    maxProfit(prices) {
        let maxProf = 0
        let buy = prices[0]

        for (let r = 0; r < prices.length; r ++) {
            maxProf = Math.max(maxProf, prices[r] - buy)

            if (prices[r] < buy) {
                buy = prices[r]
            }
        }

        return maxProf
    }
}
